rct229/rulesets/ashrae9012019/ruleset_functions/get_energy_required_to_heat_swh_use.py

```diff
@@ -0,0 +1,228 @@
+from rct229.rulesets.ashrae9012019.ruleset_functions.get_spaces_served_by_swh_use import (
+    get_spaces_served_by_swh_use,
+)
+from rct229.schema.config import ureg
+from rct229.schema.schema_enums import SchemaEnums
+from rct229.utils.assertions import assert_, getattr_
+from rct229.utils.jsonpath_utils import find_all, find_exactly_one_with_field_value
+from rct229.utils.pint_utils import ZERO
+from rct229.utils.utility_functions import (
+    find_exactly_one_schedule,
+    find_exactly_one_service_water_heating_distribution_system,
+    find_exactly_one_space,
+    find_exactly_one_service_water_heating_use,
+)
+
+SERVICE_WATER_HEATING_USE_UNIT = SchemaEnums.schema_enums[
+    "ServiceWaterHeatingUseUnitOptions"
+]
+
+REQUIRED_USE_UNIT = [
+    SERVICE_WATER_HEATING_USE_UNIT.POWER,
+    SERVICE_WATER_HEATING_USE_UNIT.VOLUME,
+]
+
+VOLUME_BASED_USE_UNIT = [
+    SERVICE_WATER_HEATING_USE_UNIT.VOLUME_PER_PERSON,
+    SERVICE_WATER_HEATING_USE_UNIT.VOLUME_PER_AREA,
+    SERVICE_WATER_HEATING_USE_UNIT.VOLUME,
+]
+
+WATER_DENSITY = 8.3452 * ureg("lb/gallon")
+WATER_SPECIFIC_HEAT = 1.001 * ureg("Btu/lb/delta_degF")
+
+
+def get_energy_required_to_heat_swh_use(
+    swh_use_id: str, rmd: dict, building_segment_id: str, is_leap_year: bool
+) -> dict[str, Quantity | None]:
+    """
+    This function calculates the total energy required to heat the SWH use over the course of a year.  Note - this function does not work for service water heating uses with use_units == "OTHER".  In this case, it will return 0 Btu.
+
+    Parameters
+    ----------
+    swh_use_id: str, id of service_water_heating_uses
+    rmd: dict, RMD at RuleSetModelDescription level
+    building_segment_id: str, id of building_segment
+    is_leap_year: bool, default: False
+        Whether the year is a leap year or not.
+
+    Returns
+    ----------
+    energy_required_by_space: A dict where the keys are space_ids and values are the total energy required to heat the swh_use for that space.  If a swh_use is not assigned to any spaces, the key will be "NO_SPACES_ASSIGNED; if the swh_use.use_units == 'OTHER', the total energy required will be set to None"
+
+    """
+    swh_use = find_exactly_one_service_water_heating_use(rmd, swh_use_id)
+    building_segment = find_exactly_one_with_field_value(
+        "$.buildings[*].building_segments[*]", "id", building_segment_id, rmd
+    )
+    hourly_schedule_id = swh_use.get("use_multiplier_schedule")
+    hourly_schedule = (
+        find_exactly_one_schedule(rmd, hourly_schedule_id)
+        if hourly_schedule_id is not None
+        else None
+    )
+
+    is_heat_recovered_by_drain = swh_use.get("is_heat_recovered_by_drain", False)
+    use_units = getattr_(swh_use, "service_water_heating_uses", "use_units")
+    distribution_system = None
+    inlet_temperature_hourly_values = None
+    drain_heat_recovery_efficiency = 0.0
+    supply_temperature = None
+    if use_units in VOLUME_BASED_USE_UNIT or is_heat_recovered_by_drain:
```

Comment:
I kept this branch as `distribution_system` is needed for both cases (see Line 100-105). @weilixu